File: dovemq-samples/src/main/java/net/dovemq/samples/basic/BasicConsumer.java
package net.dovemq.samples.basic;

import net.dovemq.api.ConnectionFactory;
import net.dovemq.api.Consumer;
import net.dovemq.api.DoveMQMessage;
import net.dovemq.api.DoveMQMessageReceiver;
import net.dovemq.api.Session;

/**
 * This sample shows how to create a DoveMQ consumer that creates a
 * queue in the DoveMQ broker, and waits for incoming messages.
 */
public class BasicConsumer {
    private static final String QUEUE_NAME = "SampleQueue";

    /**
     * Implementation of a sample MessageReceiver callback, that is registered
     * with the Consumer.
     */
    private static class SampleMessageReceiver implements DoveMQMessageReceiver {
        @Override
        public void messageReceived(DoveMQMessage message) {
            byte[] body = message.getPayload();
            String payload = new String(body);
            System.out.println("Received message: " + payload);
        }
    }

    public static void main(String[] args) {
        /*
         * Read the broker IP address passed in as -Ddovemq.broker Defaults to
         * localhost
         */
        String brokerIp = System.getProperty("dovemq.broker", "localhost");

        /*
         * Initialize the DoveMQ runtime, specifying an endpoint name.
         */
        ConnectionFactory.initialize("consumer");

        try {
            /*
             * Create an AMQP session.
             */
            final Session session = ConnectionFactory.createSession(brokerIp);
            System.out.println("created session to DoveMQ broker running at: " + brokerIp);

            /*
             * Create a consumer that binds to a queue on the broker.
             */
            Consumer consumer = session.createConsumer(QUEUE_NAME);

            /*
             * Register a message receiver with the consumer to asynchronously
             * receive messages.
             */
            SampleMessageReceiver messageReceiver = new SampleMessageReceiver();
            consumer.registerMessageReceiver(messageReceiver);

            System.out.println("waiting for messages. Press Ctl-C to shut down consumer.");
            /*
             * Register a shutdown hook to perform graceful shutdown.
             */
            Runtime.getRuntime().addShutdownHook(new Thread() {
                @Override
                public void run() {
                    /*
                     * Close the AMQP session
                     */
                    session.close();

                    /*
                     * Shutdown DoveMQ runtime.
                     */
                    ConnectionFactory.shutdown();
                }
            });
        }
        catch (Exception ex) {
            System.out.println("Caught Exception: " + ex.toString());
            /*
             * Shutdown DoveMQ runtime.
             */
            ConnectionFactory.shutdown();
        }
    }
}

File: dovemq-scala-samples/src/main/scala/net/dovemq/samples/basic/BasicConsumer.scala
package net.dovemq.samples.basic

import net.dovemq.api.{DoveMQMessageReceiver, DoveMQMessage, ConnectionFactory}

/**
 * Implementation of a sample MessageReceiver callback, that is registered
 * with the Consumer.
 */
class SampleMessageReceiver extends DoveMQMessageReceiver {
  override def messageReceived(message: DoveMQMessage) = {
    val body = message.getPayload()
    val payload = new String(body)
    println("Received message: " + payload)
  }
}

/**
 * This sample shows how to create a DoveMQ consumer that creates a transient
 * queue in the DoveMQ broker, and waits for incoming messages.
 */
object BasicConsumer {

  def main(args: Array[String]): Unit = {

    val queueName = "SampleQueue"
    val brokerIp = System.getProperty("dovemq.broker", "localhost")

    ConnectionFactory.initialize("producer")

    val session = ConnectionFactory.createSession(brokerIp)
    println("created session to DoveMQ broker running at: " + brokerIp)

    val consumer = session.createConsumer(queueName)

    /*
     * Register a message receiver with the consumer to asynchronously
     * receive messages.
     */
    val messageReceiver = new SampleMessageReceiver()
    consumer.registerMessageReceiver(messageReceiver)

    println("waiting for messages. Press Ctl-C to shut down consumer.")
    /*
     * Register a shutdown hook to perform graceful shutdown.
     */
    Runtime.getRuntime().addShutdownHook(new Thread() {
      override def run() {
        session.close()
        ConnectionFactory.shutdown()
      }
    });
  }
}